crossbar/worker/proxy.py

```diff
@@ -60,366 +60,438 @@ class ProxyWorkerProcess(worker.NativeWorkerProcess):
+                session_meta_api=True,
+                registration_meta_api=True,
+                shared_registration=True,
+                progressive_call_results=True,
+                registration_revocation=True,
+                payload_transparency=True,
+                testament_meta_api=True,
+                payload_encryption_cryptobox=True,
+                call_canceling=True,
+            )
+        }
+
+        msg = None
+        if isinstance(res, types.Accept):
+            msg = message.Welcome(
+                self._session_id,
+                roles,
+                realm=self._realm,
+                authid=self._authid,
+                authrole=self._authrole,
+                authmethod=self._authmethod,
+                authprovider=self._authprovider,
+                authextra=self._authextra,
+                #  custom=custom,
+            )
+
+        elif isinstance(res, types.Challenge):
+            msg = message.Challenge(
+                res.method,
+                res.extra,
+            )
 
-        messages = self._serializer.unserialize(payload, isBinary)
+        elif isinstance(res, types.Deny):
+            msg = message.Abort(
+                res.reason,
+                res.message,
+            )
 
-        if self._awaiting_hello:
-            if len(messages) != 1:
-                raise RuntimeError(
-                    "While waiting for Hello message, got {} WAMP "
-                    "messages (expected exactly one)".format(len(messages))
+        else:
+            msg = message.Abort("wamp.error.runtime_error", "Internal logic error")
+
+        if isinstance(msg, message.Welcome):
+
+            @inlineCallbacks
+            def _backend_connected(backend_session):
+                """
+                we have done authentication with the client; now we can connect to
+                the backend (and we wait to tell the client they're
+                welcome until we have actually connected to the
+                backend).
+                """
+
+                yield backend_session._on_connect
+                self._backend_session = backend_session
+
+                def welcome(session, details):
+                    """
+                    finally pass our Welcome to the frontend once we've Joined on the
+                    backend
+                    """
+                    self.transport.send(msg)
+                self._backend_session.on('join', welcome)
+
+                key = _read_node_key(self._controller._cbdir, private=False)
+                authmethods = [
+                    '{}-proxy'.format(x)
+                    for x in self._backend_session._authenticators.keys()
+                ]
+
+                self._backend_session.join(
+                    self._realm,
+                    authmethods=authmethods,
+                    authid=None,
+                    authrole=self._authrole,  # same role as front-end?
```

Comment:
I would argue `authrole==None` would make sense - but that it doesn't matter because it is not used in the backend?

in general, my view:

the backend should authenticate the incoming proxy connection either "anonymous" (with a fixed credential defined only by the backend) or "cryptosign" (_only_ relying on the pubkey presented by the proxy, _not_ anything requested by the proxy in authid/authrole/realm).

_if_ that authentication succeeds, then the backend will (without wamp MUX) authenticate the incoming connection using the credentials supplied in "cb_proxy_*"

in this case the realm/authid/authrole requested in HELLO in the proxy=>backend connection is never actually used - and hence should be set (or omitted) by the proxy in the first place